safenode/src/storage/registers.rs

```diff
@@ -0,0 +1,735 @@
+        // cmd let's verify it's valid before accepting it, however 'Edits cmds' cannot be
+        // verified untill we have the `Register create` cmd.
+        match (stored_reg.state.as_mut(), cmd) {
+            (Some(_), RegisterCmd::Create { .. }) => return Ok(()), // no op, since already created
+            (Some(ref mut register), RegisterCmd::Edit(_)) => self.apply(cmd, register)?,
+            (None, RegisterCmd::Create(cmd)) => {
+                // the target Register is not in our store or we don't have the 'Register create',
+                // let's verify the create cmd we received is valid and try to apply stored cmds we may have.
+                let SignedRegisterCreate { op, auth } = cmd;
+                auth.verify_authority(serialize(op)?)?;
+
+                trace!("Creating new register: {:?}", cmd.dst_address());
+                // let's do a final check, let's try to apply all cmds to it,
+                // those which are new cmds were not validated yet, so let's do it now.
+                let mut register =
+                    Register::new(*op.policy.owner(), op.name, op.tag, op.policy.clone());
+
+                for cmd in &stored_reg.op_log {
+                    self.apply(cmd, &mut register)?;
+                }
+
+                stored_reg.state = Some(register);
+            }
+            (None, _edit_cmd) => { /* we cannot validate it right now, but we'll store it */ }
+        }
+
+        stored_reg.op_log.push(cmd.clone());
+        Ok(())
+    }
+
+    // Try to apply the provided cmd to the register state, performing all op validations
+    fn apply(&self, cmd: &RegisterCmd, register: &mut Register) -> Result<()> {
+        let addr = cmd.dst_address();
+        if &addr != register.address() {
+            return Err(Error::RegisterAddrMismatch {
+                cmd_dst_addr: addr,
+                reg_addr: *register.address(),
+            });
+        }
+
+        match cmd {
+            RegisterCmd::Create { .. } => Ok(()),
+            RegisterCmd::Edit(SignedRegisterEdit { op, auth }) => {
+                auth.verify_authority(serialize(op)?)?;
+
+                info!("Editing Register: {addr:?}");
+                let public_key = auth.public_key;
+                register.check_permissions(Action::Write, Some(User::Key(public_key)))?;
+                let result = register
+                    .apply_op(op.edit.clone())
+                    .map_err(Error::NetworkData);
+
+                match result {
+                    Ok(()) => {
+                        trace!("Editing Register success: {addr:?}");
+                        Ok(())
+                    }
+                    Err(err) => {
+                        trace!("Editing Register failed {addr:?}: {err:?}");
+                        Err(err)
+                    }
+                }
+            }
+        }
+    }
+
+    // Gets stored register log from disk, trying to reconstruct the Register
```

Comment:
The docs say from disk here. Absolute nit pick, but I guess as it's now seen, we might just as well update it.